### PR TITLE
CGE-182: Created AdvertisementDisplayDto

### PR DESCRIPTION
**Pull request by Jack11M**
Created Advertisement Display Dto.
Mapped it to findById method in Advertisement Service.

### CODE REVIEW

File: src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java [hunk 1]
```diff
@@ -199,4 +217,47 @@ private AdvertisementTitleDto buildAdvertisementTitle(Advertisement advertisemen
...
+        return advertisement.getUser().getPhones().stream()
+                .findFirst()
+                .map(phone -> String.valueOf(phone.getPhoneNumber()))
+                .orElse("");
+    }
+
+    private String getOwnerFullName(Advertisement advertisement) {
+        User owner = advertisement.getUser();
+        String firstName = Optional.of(owner.getFirstName()).orElse("");
+        String lastName = Optional.of(owner.getLastName()).orElse("");
```
Comment (svyatoslav145) [inline src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java:259]: would you like to put a space here?
String ownerName = String.format("%s%s%s", firstName()," " ,lastName()).trim();

**Review by svyatoslav145 — APPROVED** [review]

File: src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java (hunk 1 above)
Comment (Jack11M) [reply src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java:259]: I have added a space in a different way.
Why your approach is better?

Comment (Jack11M) [reply src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java:259]: String.format("%s %s", firstName(), "lastName()).trim();
We already have space here between first name and last name.
You can view it in the test case.

Comment (svyatoslav145) [reply src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java:259]: not only did not see, but was stupid

File: src/main/java/com/hillel/items_exchange/controller/AdvertisementController.java [hunk 2]
```diff
@@ -160,15 +160,13 @@ public AdvertisementDto updateAdvertisement(@Validated(Exist.class) @RequestBody
             @ApiResponse(code = 200, message = "OK"),
             @ApiResponse(code = 400, message = "BAD REQUEST"),
             @ApiResponse(code = 403, message = "FORBIDDEN")})
-    public ResponseEntity<HttpStatus> deleteAdvertisement(@Valid @RequestBody AdvertisementDto dto, Principal principal)
+    public ResponseEntity<HttpStatus> deleteAdvertisement(@Valid @RequestBody Long advertisementId, Principal principal)
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/controller/AdvertisementController.java:163]: Thanks 4 that! 👍 
But you need to check if Swagger displays that parameter and describe it.

File: src/main/java/com/hillel/items_exchange/controller/AdvertisementController.java [hunk 3]
```diff
@@ -160,15 +160,13 @@ public AdvertisementDto updateAdvertisement(@Validated(Exist.class) @RequestBody
...
             throws IllegalOperationException {
 
         User owner = getUser(principal.getName());
-        validateAdvertisementOwner(dto, owner);
-        Optional<AdvertisementDto> byId = advertisementService.findDtoById(dto.getId());
-
-        if (byId.isPresent() && byId.get().equals(dto)) {
-            advertisementService.remove(dto.getId());
+        validateAdvertisementOwner(advertisementId, owner);
+        if (advertisementService.existById(advertisementId)) {
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/controller/AdvertisementController.java:168]: Why do we need that? In line above we checked that it exists.

File: src/main/java/com/hillel/items_exchange/controller/AdvertisementController.java [hunk 4]
```diff
@@ -160,15 +160,13 @@ public AdvertisementDto updateAdvertisement(@Validated(Exist.class) @RequestBody
...
+            advertisementService.remove(advertisementId);
             return ResponseEntity.status(HttpStatus.OK).build();
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/controller/AdvertisementController.java:170]: Please response with annotation.

File: src/main/java/com/hillel/items_exchange/dto/AdvertisementDisplayDto.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    private DealType dealType;
+    private byte[] ownerAvatar;
+    private String createdDate;
+    private String description;
+    private LocationDto location;
+    private List<ImageDto> images;
+    private boolean readyForOffers;
+    private String wishesToExchange;
+    private CategoryNameDto category;
+    private SubcategoryDto subcategory;
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/dto/AdvertisementDisplayDto.java:45]: You're real pedant doing that! 🌲

File: src/main/java/com/hillel/items_exchange/dto/AdvertisementDisplayDto.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+import com.hillel.items_exchange.model.enums.DealType;
+import com.hillel.items_exchange.model.enums.Gender;
+import com.hillel.items_exchange.model.enums.Season;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@EqualsAndHashCode(exclude = {"advertisementId", "ownerAvatar", "category", "location", "images"})
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/dto/AdvertisementDisplayDto.java:24]: Do we really use equals and hashcode for that dto?

File: src/main/java/com/hillel/items_exchange/dto/CategoryNameDto.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+import com.hillel.items_exchange.annotation.Zero;
+import com.hillel.items_exchange.mapper.transfer.New;
+import io.swagger.annotations.ApiModelProperty;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Getter
+@Setter
+@ToString
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/dto/CategoryNameDto.java:21]: Where do we need that?

File: src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java [hunk 8]
```diff
@@ -199,4 +217,47 @@ private AdvertisementTitleDto buildAdvertisementTitle(Advertisement advertisemen
                 .ownerAvatar(advertisement.getUser().getAvatarImage())
                 .build();
     }
+
+    private AdvertisementDisplayDto buildAdvertisementDisplayDto(Advertisement advertisement) {
+        Location advLocation = advertisement.getLocation();
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java:222]: I couldn't found where you use that variable...

File: src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java [hunk 9]
```diff
@@ -199,4 +217,47 @@ private AdvertisementTitleDto buildAdvertisementTitle(Advertisement advertisemen
...
+        long id = advertisement.getId();
+        Subcategory subcategory = advertisement.getSubcategory();
+        String createdDate = advertisement.getCreated().format(DateTimeFormatter.ofPattern(dateFormat));
+        String age = Optional.ofNullable(advertisement.getAge()).map(AgeRange::getValue).orElse("");
+        return AdvertisementDisplayDto.builder()
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java:227]: As for me, you could use BeanUtils to copy main part of parameters and use builder for specific ones. What do you think about it?

File: src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java [hunk 10]
```diff
@@ -199,4 +217,47 @@ private AdvertisementTitleDto buildAdvertisementTitle(Advertisement advertisemen
...
+                .subcategory(convertTo(subcategory, SubcategoryDto.class))
+                .category(convertTo(subcategory.getCategory(), CategoryNameDto.class))
+                .createdDate(createdDate)
+                .build();
+    }
+
+    private String getOwnerPhone(Advertisement advertisement) {
+        return advertisement.getUser().getPhones().stream()
+                .findFirst()
+                .map(phone -> String.valueOf(phone.getPhoneNumber()))
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java:252]: Just pass User here and make it much simpler:
```
var phones = user.getPhones();
if(!phones.isEmpty())
    return String.valueOf(phones.get(0));
return "";
```

File: src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java (hunk 1 above)
Comment (rpkyrych) [reply src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java:259]: Guys, why so complicated???
```
private String getOwnerFullName(User user) {
        String formatted = String.format("%s %s", user.getFirstName(), user.getLastName());
        return formatted.isBlank() ? user.getUsername() : formatted;
    }
```

**Review by rpkyrych — APPROVED** [review]
Looks good.

File: src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java (hunk 1 above)
Comment (Jack11M) [reply src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java]: Thank you. Changed.

File: src/main/java/com/hillel/items_exchange/controller/AdvertisementController.java (hunk 3 above)
Comment (Jack11M) [reply src/main/java/com/hillel/items_exchange/controller/AdvertisementController.java]: Removed.

File: src/main/java/com/hillel/items_exchange/controller/AdvertisementController.java (hunk 2 above)
Comment (Jack11M) [reply src/main/java/com/hillel/items_exchange/controller/AdvertisementController.java]: Ok

File: src/main/java/com/hillel/items_exchange/controller/AdvertisementController.java (hunk 4 above)
Comment (Jack11M) [reply src/main/java/com/hillel/items_exchange/controller/AdvertisementController.java]: Done

File: src/main/java/com/hillel/items_exchange/dto/AdvertisementDisplayDto.java (hunk 6 above)
Comment (Jack11M) [reply src/main/java/com/hillel/items_exchange/dto/AdvertisementDisplayDto.java]: Removed

File: src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java (hunk 8 above)
Comment (Jack11M) [reply src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java]: Changed

File: src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java (hunk 9 above)
Comment (Jack11M) [reply src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java]: Great idea!

File: src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java (hunk 10 above)
Comment (Jack11M) [reply src/main/java/com/hillel/items_exchange/service/impl/AdvertisementServiceImpl.java:242]: Changed

File: src/main/java/com/hillel/items_exchange/dto/CategoryNameDto.java (hunk 7 above)
Comment (Jack11M) [reply src/main/java/com/hillel/items_exchange/dto/CategoryNameDto.java]: Removed

**Review by drw85 — APPROVED** [review]

**Pull request merged**